provider/kubeletconfig/kubeletconfig_resource.go

```diff
@@ -0,0 +1,261 @@
+
+	resp.Schema = schema.Schema{
+		Description: "KubeletConfig Configuration",
+		Attributes: map[string]schema.Attribute{
+			"cluster": schema.StringAttribute{
+				Required:    true,
+				Description: "Identifier of the cluster.",
+				PlanModifiers: []planmodifier.String{
+					common.Immutable(),
+				},
+			},
+			"pod_pids_limit": schema.Int64Attribute{
+				Required: true,
+				Description: "Sets the requested podPidsLimit to be applied as part of the custom " +
+					"KubeletConfig.",
+				Validators: []validator.Int64{
+					PidsLimitValidator{},
+				},
+			},
+		},
+	}
+}
+
+func (k *KubeletConfigResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	state := k.getKubeletConfigStateFromPlan(ctx, req.Plan, &resp.Diagnostics)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	clusterId := state.Cluster.ValueString()
+	err := k.clusterWait.WaitForClusterToBeReady(ctx, clusterId)
+
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Cluster is not ready",
+			fmt.Sprintf("Cluster with id '%s' is not in the ready state: %v", clusterId, err),
+		)
+		return
+	}
+
+	exists, _, err := k.configClient.Exists(ctx, clusterId)
+	if err != nil || exists {
+		resp.Diagnostics.AddError(failedToCreateSummary,
+			fmt.Sprintf("KubeletConfig for cluster '%s' may already exist: %v", clusterId, err))
+		return
+	}
+
+	kubeletConfig, err := k.convertStateToApiResource(state)
+	if err != nil {
+		resp.Diagnostics.AddError(failedToCreateSummary,
+			fmt.Sprintf("Failed to build request to create KubeletConfig on cluster '%s': %v", clusterId, err))
+		return
+	}
+
+	_, err = k.configClient.Create(ctx, clusterId, kubeletConfig)
+	if err != nil {
+		resp.Diagnostics.AddError(failedToCreateSummary,
+			fmt.Sprintf("Failed to create KubeletConfig on cluster '%s': %v", clusterId, err))
+		return
+	}
+
+	k.writeStateToResponse(ctx, state, &resp.State, &resp.Diagnostics)
+}
+
+func (k *KubeletConfigResource) convertStateToApiResource(state *KubeletConfigState) (*cmv1.KubeletConfig, error) {
+	builder := cmv1.KubeletConfigBuilder{}
+	builder.PodPidsLimit(int(state.PodPidsLimit.ValueInt64()))
+	return builder.Build()
+}
+
+func (k *KubeletConfigResource) convertApiResourceToState(
+	kubeletConfig *cmv1.KubeletConfig, state *KubeletConfigState) {
+	state.PodPidsLimit = types.Int64Value(int64(kubeletConfig.PodPidsLimit()))
+}
+
+func (k *KubeletConfigResource) getKubeletConfigStateFromState(
+	ctx context.Context, state tfsdk.State, diags *diag.Diagnostics) *KubeletConfigState {
+	ks := &KubeletConfigState{}
+	ds := state.Get(ctx, ks)
+	diags.Append(ds...)
+	return ks
+}
+
+func (k *KubeletConfigResource) getKubeletConfigStateFromPlan(
+	ctx context.Context, plan tfsdk.Plan, diags *diag.Diagnostics) *KubeletConfigState {
+	ks := &KubeletConfigState{}
+	ds := plan.Get(ctx, ks)
+	diags.Append(ds...)
+	return ks
+}
```

Comment:
Those two functions are duplication, lets keep one with name `getKubeletConfigState`

Reply:
@nirarg One function returns the state from `tfsdk.State` and the other from `tfsdk.Plan`. The functions that return the state using the `tfsdk.Plan` are called by the `delete` and `read` methods of the resource, which AFAICS, don't have access to the `tfsdk.State` struct.

Am I missing your point?

Reply:
Oh, I see it now, my mistake ... resolved :-)